doc/spec/debug-ros-nodes.md

```diff
@@ -0,0 +1,269 @@
+| `gdb -ex run --args` | run node in `gdb` in the same `xterm` |
+| `valgrind` | run node in `valgrind` |
+
+Currently, this approach only provides [limited](#why-no-launch-prefix-support-for-rospy) debugger support to `roscpp` nodes.
+
+## Goals
+
+* enable attaching into a local C++ or Python process launched through `roslaunch` to debug that process
+* enable debugging C++ and Python nodes specified in a `.launch` file at entry
+
+### Non-goals
+
+Language-specific debugging capabilities including:
+
+* complicated debugging scenarios like debugging a remote process
+
+We would want to utilize existing language-specific tools for [C++][ms-vscode.cpptools] and [Python][ms-python.python]:
+
+![architecture][architecture]
+
+## Debugging in Visual Studio Code
+
+Typically, to start a debugging sessions, a debug configuration should be created in `launch.json` first.
+After that, the newly created configuration could be selected and executed by pressing the green arrow or the `F5` key.
+
+![execute_a_debug_configuration][execute_a_debug_configuration]
+
+What happens afterwards gets handled by 2 processes:
+
+* the running Visual Studio Code instance (this is the running context of all extensions)
+* the debug adapter instance (for each language)
+
+A debug configuration first gets resolved by a registered debug configuration provider in VS Code's context.
+Afterwards, a debug request gets sent to the extension’s debug adapter, which actually handles the debugging process.
+However, the debug adapter is a separate process from the VS Code process, so no `vscode` APIs could be called from there.
+
+In our scenario, we need to do everything in the extension context to utilize the functionalities provided by other extensions.
+Ideally, the workflow should be as simple as possible:
+
+1. the user creates a `ros`-type debug configuration
+2. our extension resolves the configuration
+3. our extension generates debug configuration for C++ or/and Python extensions
+
+![debug_flow][debug_flow]
+
+As illustrated in the above flow chart, we would also [need a stub debug adapter](#why-is-a-stub-debug-adapter-needed).
+
+## Attach
+
+Our goal for attach-debug is to provide a way to debug a ROS node while not changing anything in an existing flow.
+This means when the user executes a `.launch` file with `roslaunch` just like before, he/she could use `vscode-ros` to easily start an attach-debug session to debug a running ROS node process.
+All attach-debug configurations are propagated to language-specific extensions.
+
+![attach_debug][attach_debug]
+
+### Attach into a C++ process
+
+Just like using Visual Studio or windbg, attaching into a running C++ process could be achieved with `vscode-cpptools` easily.
+Our extension automatically chooses `cppvsdbg` for MSVC debugger when it's running on Windows, and `cppdbg` for gdb-based debugger otherwise.
+
+### Attach into a Python process
+
+Our extension uses [`ptvsd`][ptvsd] to enable debugging Python scripts.
```

Comment:
We should also mention how we use `getRemoteLauncherCommand` API in Python extension for this purpose and what's encapsulated by that extension.

Reply:
adding reference to [vscode-python/api.ts](https://github.com/microsoft/vscode-python/blob/master/src/client/api.ts), but since ultimately we want to use an updated API interface (instead of hacking `getRemoteLauncherCommand`), not including the implementation detail here.